artiq/firmware/libboard_artiq/drtio_eem.rs

```diff
@@ -0,0 +1,195 @@
+        unsafe {
+            core::slice::from_raw_parts(
+                (self as *const SerdesConfig) as *const u8,
+                core::mem::size_of::<SerdesConfig>(),
+            )
+        }
+    }
+}
+
+pub fn align_eem() -> SerdesConfig {
+    let mut delay_confs: [(u8, u8, u8, u8); 4] = Default::default();
+
+    for eem_pair_no in 0..4 {
+        delay_confs[eem_pair_no] = unsafe { align(eem_pair_no) };
+    }
+
+    let mut select_odd = 0;
+    let mut decoder_invert = 0;
+    let mut delay = [0; 4];
+    let mut bitslip = [0; 4];
+    for (eem_pair_no, (select_odd_idx, dly, slip, invert)) in delay_confs.iter().enumerate() {
+        select_odd |= (select_odd_idx << eem_pair_no);
+        delay[eem_pair_no] = *dly;
+        bitslip[eem_pair_no] = *slip;
+        decoder_invert |= (invert << eem_pair_no);
+    }
+
+    SerdesConfig {
+        select_odd,
+        decoder_invert,
+        delay,
+        bitslip,
+    }
+}
+
+fn select_eem_pair(eem_pair_no: usize) {
+    unsafe {
+        csr::eem_transceiver::serdes_eem_sel_write(eem_pair_no as u8);
+    }
+}
+
+fn update_select_odd(eem_pair_no: usize, select_odd: usize) {
+    let mut odd_sel_reg = unsafe { csr::eem_transceiver::serdes_select_odd_read() };
+    // Clear bit
+    odd_sel_reg &= (!(1 << eem_pair_no));
+    // Set bit if applicable
+    unsafe { csr::eem_transceiver::serdes_select_odd_write(odd_sel_reg | (select_odd << eem_pair_no) as u8) };
+}
+
+fn update_invert(eem_pair_no: usize, invert: usize) {
+    let mut invert_reg = unsafe { csr::eem_transceiver::serdes_decoder_dly_read() };
+    // Clear bit
+    invert_reg &= (!(1 << eem_pair_no));
+    // Set bit if applicable
+    unsafe { csr::eem_transceiver::serdes_decoder_dly_write(invert_reg | (invert << eem_pair_no) as u8) };
+}
+
+fn apply_bitslip() {
+    unsafe {
+        csr::eem_transceiver::serdes_bitslip_write(1);
+        csr::eem_transceiver::serdes_bitslip_write(1);
+    }
+}
+
+fn apply_delay(tap: u8) {
+    unsafe {
+        csr::eem_transceiver::serdes_dly_cnt_in_write(tap);
+        // Ensure dly_cnt_in is updated before ld
+        clock::spin_us(100);
+        assert!(tap == csr::eem_transceiver::serdes_dly_cnt_in_read());
+        csr::eem_transceiver::serdes_dly_ld_write(1);
+    }
+}
+
+pub fn write_config(config: &SerdesConfig) {
+    unsafe {
+        csr::eem_transceiver::serdes_decoder_dly_write(config.decoder_invert as u8);
+        csr::eem_transceiver::serdes_select_odd_write(config.select_odd as u8);
+    }
+
+    for eem_pair_no in 0..4 {
+        select_eem_pair(eem_pair_no);
+        apply_delay(config.delay[eem_pair_no]);
+
+        for _ in 0..config.bitslip[eem_pair_no] {
+            apply_bitslip();
+        }
+    }
+}
+
+unsafe fn align(eem_pair_no: usize) -> (u8, u8, u8, u8) {
+    let mut table: [[bool; 32]; 20] = [[false; 32]; 20];
+
+    select_eem_pair(eem_pair_no);
+    clock::spin_us(1);
+
+    let scan = |table: &mut[[bool; 32]]| {
+        for slip in 0..5 {
```

Comment:
But ideally we should select a waveform with only one rising edge or falling edge per 8ns cycle right? Otherwise we cannot be sure if this produces the correct delay taps & bitslip.
So just guaranteeing >1UI wouldn't be enough.

Reply:
We also shouldn't need the option to select odd-indices or even-indices bits to the decoder if we only need to find an acceptable crossover point. IDELAYE2 already covers 2.5ns, we can expect another suitable crossover point approximately 20 taps away.